api/bucketsd/service.go

```diff
@@ -1497,6 +1518,367 @@ func (s *Service) PushPath(server pb.APIService_PushPathServer) (err error) {
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	req, err := server.Recv()
+	if err != nil {
+		return fmt.Errorf("on receive: %v", err)
+	}
+	var buckKey, buckRoot string
+	switch payload := req.Payload.(type) {
+	case *pb.PushPathsRequest_Header_:
+		buckKey = payload.Header.Key
+		buckRoot = payload.Header.Root
+	default:
+		return fmt.Errorf("push bucket path header is required")
+	}
+
+	lck := s.Semaphores.Get(buckLock(buckKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	buck := &tdb.Bucket{}
+	err = s.Buckets.GetSafe(ctx, dbID, buckKey, buck, tdb.WithToken(dbToken))
+	if err != nil {
+		return fmt.Errorf("getting bucket: %v", err)
+	}
+	if buckRoot != "" && buckRoot != buck.Path {
+		return status.Error(codes.FailedPrecondition, buckets.ErrNonFastForward.Error())
+	}
+
+	var wg sync.WaitGroup
+	addedCh := make(chan addedFile)
+	doneCh := make(chan struct{})
+	errCh := make(chan error)
+	go func() {
+		queue := newFileQueue()
+		for {
+			req, err := server.Recv()
+			if err == io.EOF {
+				wg.Wait() // Request ended normally, wait for pending jobs
+				close(doneCh)
+				return
+			} else if err != nil {
+				errCh <- fmt.Errorf("on receive: %v", err)
+				return
+			}
+			switch payload := req.Payload.(type) {
+			case *pb.PushPathsRequest_Chunk_:
+				pth, err := parsePath(payload.Chunk.Path)
+				if err != nil {
+					errCh <- fmt.Errorf("parsing path: %v", err)
+					return
+				}
+				fa, err := queue.add(ctx, s.IPFSClient.Unixfs(), pth, func() ([]byte, error) {
+					buck.UpdatedAt = time.Now().UnixNano()
+					buck.SetMetadataAtPath(pth, tdb.Metadata{
+						UpdatedAt: buck.UpdatedAt,
+					})
+					buck.UnsetMetadataWithPrefix(pth + "/")
+					if err = s.Buckets.Verify(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
+						return nil, fmt.Errorf("verifying bucket update: %v", err)
+					}
+					key, err := buck.GetFileEncryptionKeyForPath(pth)
+					if err != nil {
+						return nil, fmt.Errorf("getting bucket key: %v", err)
+					}
+					wg.Add(1)
+					return key, nil
+				}, addedCh, errCh)
+				if err != nil {
+					errCh <- fmt.Errorf("enqueueing file: %v", err)
+					return
+				}
+
+				if len(payload.Chunk.Data) > 0 {
+					if _, err := fa.writer.Write(payload.Chunk.Data); err != nil {
+						errCh <- fmt.Errorf("writing chunk: %v", err)
+						return
+					}
+				} else {
+					if err := fa.writer.Close(); err != nil {
+						errCh <- fmt.Errorf("closing writer: %v", err)
+						return
+					}
+				}
+			default:
+				errCh <- fmt.Errorf("invalid request")
+				return
+			}
+		}
+	}()
+
+	var changed bool
+	sctx := util.NewClonedContext(ctx)
+	saveWithErr := func(err error) error {
+		cancel()
+		if !changed {
+			return err
+		}
+		if serr := s.Buckets.Save(sctx, dbID, buck, tdb.WithToken(dbToken)); serr != nil {
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("saving bucket: %v", serr)
+		}
+		go s.IPNSManager.Publish(path.New(buck.Path), buck.Key)
+		return err
+	}
+
+	for {
```

Comment:
Here we block, waiting for... 
- The result of files added to IPFS
- Errors (if any errors occurs, the server bails)
- A final done event, which caused the server to bail w/o error

Reply:
Note that in contrast to how files _can_ be added to IPDS in "parallel", the step of updating the root dag's links and pins must happen serially. If any of the added files makes its way through this last step, the bucket _must_ be saved, _even if an error occurred elsewhere_.